src/silo/query_engine/query_result.cpp

```diff
@@ -1,16 +1,86 @@
 #include "silo/query_engine/query_result.h"
 
+#include <spdlog/spdlog.h>
+#include <iostream>
 #include <nlohmann/json.hpp>
 
 #include "silo_api/variant_json_serializer.h"
 
 namespace silo::query_engine {
 
-// NOLINTNEXTLINE(readability-identifier-naming)
-void to_json(nlohmann::json& json, const QueryResult& query_result) {
-   json = nlohmann::json{
-      {"queryResult", query_result.query_result},
-   };
+void QueryResult::clear() {
+   query_result_chunk_.clear();
+   get_chunk_ = [](std::vector<QueryResultEntry>& /*query_result_chunk*/) {};
+   i_ = 0;
+}
+
+std::optional<std::reference_wrapper<const QueryResultEntry>> QueryResult::next() {
+   auto siz = query_result_chunk_.size();
+   SPDLOG_DEBUG(
+      "next called, i = {}, is_materialized_ = {}, chunk size() = {}", i_, is_materialized_, siz
+   );
+   if (i_ >= siz) {
+      SPDLOG_DEBUG("reached the end of last chunk");
+      query_result_chunk_.clear();
+      SPDLOG_DEBUG("cleared vector");
+      get_chunk_(query_result_chunk_);
+      i_ = 0;
+      siz = query_result_chunk_.size();
+      SPDLOG_DEBUG("returned from get_chunk_, chunk size() = {}", siz);
+      if (siz == 0) {
+         SPDLOG_DEBUG("returning {} from next");
+         return {};
+      }
+   }
+   const QueryResultEntry& ref = query_result_chunk_[i_];
+   ++i_;  // can't overflow because i_ >= siz will always be true before.
+   SPDLOG_DEBUG("returning ref from next");
+   return {std::cref(ref)};
+}
+
+void QueryResult::materialize() {
+   if (is_materialized_) {
+      return;
+   }
+
+   std::vector<QueryResultEntry> tmp{};
+   while (true) {
+      get_chunk_(tmp);
+      if (tmp.empty()) {
+         break;
+      }
+      const bool optimize = true;
```

Comment:
Only if suspecting or discovering that the optimization is buggy. Or if someone would like to simplify the code and throw away the optimization, or otherwise understand why that if branch is there.

Reply:
So this is a "manual switch" in the code? Do we want to keep it or should we simply delete it?

Reply:
Yes, it is manual, and was meant to indicate that this is an optimization, that could be removed easily by toggling that boolean (or by removing the code that the boolean switches). If you don't like it I'm fine with removing it.

Reply:
If you think we might need it again at some point, we can keep it. Otherwise, I'd prefer to remove it.

Reply:
Changed into a comment.